gguf-py/examples/dump_gguf.py

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env python3
+import sys
+from pathlib import Path
+
+# Necessary to load the local gguf package
+sys.path.insert(0, str(Path(__file__).parent.parent))
+
+from gguf import GGUFReader, GGUFValueType  # noqa: E402
+
+
+# For more information about what field.parts and field.data represent,
+# please see the comments in the modify_gguf.py example.
+def dump_gguf(filename: str) -> None:
+    print(f'* Loading: {filename}')
+    reader = GGUFReader(filename, 'r')
+    print(f'\n* Dumping {len(reader.fields)} key/value pair(s)')
+    for n, field in enumerate(reader.fields.values(), 1):
+        if not field.types:
+            pretty_type = 'N/A'
+        elif field.types[0] == GGUFValueType.ARRAY:
+            nest_count = len(field.types) - 1
+            pretty_type = '[' * nest_count + str(field.types[-1].name) + ']' * nest_count
+        else:
+            pretty_type = str(field.types[-1].name)
+        print(f'  {n:5}: {pretty_type:10} | {len(field.data):8} | {field.name}', end = '')
+        if len(field.types) == 1:
+            curr_type = field.types[0]
+            if curr_type == GGUFValueType.STRING:
+                print(' = {0}'.format(repr(str(bytes(field.parts[-1]), encoding='utf8')[:60])), end = '')
+            elif field.types[0] in reader.gguf_scalar_to_np:
+                print(' = {0}'.format(field.parts[-1][0]), end = '')
+        print()
+
+    print(f'\n* Dumping {len(reader.tensors)} tensor(s)')
+    for n, tensor in enumerate(reader.tensors, 1):
+
+        prettydims = ', '.join('{0:5}'.format(d) for d in list(tensor.shape) + [1] * (4 - len(tensor.shape)))
+        print(f'  {n:5}: {tensor.n_elements:10} | {prettydims} | {tensor.tensor_type.name:7} | {tensor.name}')
+
+
+if __name__ == '__main__':
+    if len(sys.argv) < 2:
```

Comment:
What about using `argparse` to work with CLI args --it gives help message and argument checking for free. Using `sys.args` directly is not a good practice in Python most of the time.


And if we expect these executables to serve users as an actual tool rather than simply an example to demonstrate the usage, we can install them as entry points in `pyproject.toml`, so users can invoke them easily on the command line after pip-installing.

Reply:
Thanks for the comment!

Sure, that's a reasonable suggestion and won't make those examples much more complicated. I was trying to hit a balance between usable tool and an example that was still simple enough to be helpful (I actually think I went too far trying to make the output pretty).

Also, for `llama.cpp` people, something hidden away in `gguf-py/examples` isn't really going to be that discoverable. So I'm not 100% sure where those should be. Should the endian converter script also be there?

Reply:
Yes, I think argparse can make it even more readable and intuitive for Python people.

For the entry point side, they look to me more like simple but helpful utilities rather than API examples --except for the writing example.

I think those ones except for the writing script can be moved to `scripts` folder and then installed in `pyproject.toml`.

Reply:
We probably should use a more consistent naming convention if it's getting installed into the user's path. How about something like this (all under `gguf-py/scripts/`):

1. `gguf-dump.py`
2. `gguf-set-kv.py`
3. `gguf-convert-endian.py`

I don't really have strong feelings about the name so if anyone else wants something different, that's probably not an issue. I do think they should start with a common prefix though.

Reply:
Prefixing with `gguf-` looks like a good idea. So they can run in two ways:

1. When not installed to the environment: `python path/to/gguf-py/scripts/gguf-dump` (adding Python because shabang is not universally supported)
2. When installed with pip: `gguf-dump`

For the latter, we need to add them in `[tool.poetry.scripts]` in `pyproject.toml`. My clip.cpp Python binding is an [example](https://github.com/monatis/clip.cpp/blob/main/examples/python_bindings/pyproject.toml) (the very end of the file).

Reply:
@monatis I updated the scripts to use `argparse` and reorganized stuff a bit. It didn't see clear how to set up the Poetry thing though.

I want it to be possible to run the scripts like:

    python gguf-py/scripts/gguf-dump.py blah.gguf

And that should work without having to install anything, it should just automatically use the repo-local `gguf` package in that case.

It seems like Poetry wants all the scripts to be in a package that doesn't run standalone?

Reply:
We're not limited to a single package --we can add a `__init__.py` file to `scripts` and then include `scripts` in `packages` together with `gguf` in `pyproject.toml`:

```
packages = [
    {include = "gguf"},
    {include = "gguf/py.typed"},
    {include = "scripts"},
]
```

 Then we can list them one by one as in the following:

```
[tool.poetry.scripts]
gguf-dump = 'scripts.gguf_dump'
```

Not tested but such a setup should work.

Reply:
@monatis 

After much struggling because it can't really handle files with `-` in the name, this is what I came up with. In `pyproject.toml`:

```toml
[tool.poetry.scripts]
gguf-convert-endian = "scripts:gguf_convert_endian_entrypoint"
gguf-dump = "scripts:gguf_dump_entrypoint"
gguf-set-metadata = "scripts:gguf_set_metadata_entrypoint"
```

In `scripts/__init__.py`:

```python
import importlib

gguf_convert_endian_entrypoint = importlib.import_module("scripts.gguf-convert-endian").main
gguf_dump_entrypoint           = importlib.import_module("scripts.gguf-dump").main
gguf_set_metadata_entrypoint   = importlib.import_module("scripts.gguf-set-metadata").main
```

Before I go ahead and commit that, can you live with that approach? I'm not gonna rename my scripts just to make poetry happy, dammit. :)

Reply:
ok to me :)